### DISCUSSION TITLE
uid undefined and null - due to multiple users

### DISCUSSION THREAD

**Discussion by cmeffley** [nss-evening-cohort-14/evening-client-side#285] [Capstones!] [answered]
Chris Meffley

Did not see any tickets that I thought would help.

When trying to add or update a race, I get the error message: TypeError: Cannot read property 'uid' of undefined (when editing)
and TypeError: Cannot read property 'uid' of null (when adding). The issue is coming from my races object that has both the coach and athlete uid in it. If I'm logged in under coach, the the error comes from the athlete uid in the object, and if I'm logged in as athlete, the error comes from the coach uid in the object.

https://stackoverflow.com/questions/64159970/why-is-currentuser-uid-undefined
https://stackoverflow.com/questions/29836168/javascript-check-for-multiple-undefined-objects
https://stackoverflow.com/questions/65469282/firebase-data-not-defined-when-used-inside-react-usestate
Struggling to find relative documents about this - still searching.

Tried:
In the object:
- `athleteUid: athlete.uid`
- `athleteUid: athlete.uid || null`
- `athleteUid: athlete.uid || ' ' `
- `athleteUid: athlete.uid || coach.uid`
- The same things for coach as well

Files:
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/components/RacesForm.js
https://github.com/cmeffley/front-end-capstone/tree/crudraces/src/components
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/views/RacesFormView.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/views/RaceScheduleView.js

**Comment by drteresavasquez** [comment]
Can I get a screen shot of your Firebase nodes for these?

**Comment by cmeffley** [comment, marked as answer]
![2021-06-12](https://user-images.githubusercontent.com/76795299/121781503-15104f00-cb73-11eb-862c-4adca6086516.png)
I need to add an athlete node in order to be able to select the specific athlete associated with the race (vice versa for coach).
Then I need to make a call to firebase when logging in to make sure the user does or does not exist. Once the users exists, I'll use the user object with their uid to set the user (athlete or coach), which I can use in the form as a drop down so the uid is then attached to the race.

**Comment by cmeffley** [comment]
Update:
Now getting TypeError: Cannot read property 'athleteUid' of null in my race view. 
console.warn shows correct responses for:
-  response.data in getSingleAthlete
- response in index.js where calling getSingleAthlete
But, when console.warn athlete, it renders null, so setAthlete is not updating which should be causing the error message.

Searched:
https://stackoverflow.com/questions/36085726/why-is-setstate-in-reactjs-async-instead-of-sync/36087156#36087156
https://stackoverflow.com/questions/41446560/react-setstate-not-updating-state
https://stackoverflow.com/questions/38558200/react-setstate-not-updating-immediately

Files:
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/App/index.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/views/RacesFormView.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/helpers/data/athleteData.js

**Reply by drteresavasquez** [reply]
@cmeffley is this what we worked through at the end of class?

**Reply by cmeffley** [reply]
This is a new bug that popped up right after we fixed what we went over in class.  The race cards are not rendering for the athlete that is logged in (getting the above error message). To me, it looks like setAthlete is not updating at the end of the getSingleAthlete call in index.js. But I could be missing something else.
